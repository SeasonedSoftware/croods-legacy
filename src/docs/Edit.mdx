---
name: Edit
route: /edit
order: 100
---

# Edit

The Edit component is a union of Info and New.
This component firstly does a `GET` request to the `path` like Info. Then its update function, when executed, does a `PUT` request with the JSONified data, like the New component.


## API

|Property|Type|Required|Default|Description|
|--- |--- |--- |--- |--- |
|id|String/Number|true|-|Ex: 1234 or '1234'|
|name|String|true|-|Defines the deep properties used in the component. Ex: foo.bar|
|render|Func|true|-|A function returning a React Node. Ex: ({ info = {}, update = (id, ...attributes), updating = bool, error }, props) -> < JSX />|
|renderUpdated|Func|false|-|A function returning a React Node. Ex: (updated = {}) -> < JSX />|
|parentId|String/Number|false|-|Parent element's Id, for more specificity in changes.|
|path|String|false|`/${name}/${id}`|Ex: /foo/bar-_12?&=|
|parseInfoResponse|Func|false|-|Parse specific responses to adjust the API to croods patterns. Ex: (json, response, requestAttributes) -> Object|
|parseUpdateResponse|Func|false|-||

## How to use

You have to add a reducer with the same name that you will pass to Edit.

```javascript
import { createReducer } from 'croods'

const reducers = {
  foobar: createReducer('foobar'),
}
```

### render

The render function receives four parameters: **info, update, updating and error**.

- info: the response body of the `GET` request.
- update: the function to make the `PUT` request.
- updating: a boolean that identifies when the `PUT` request is being done and it is not finished.
- error: The error message in case the update fails.

### renderUpdated

In order to do something after the `update`call, use renderUpdated function to
 redirect to other page or show the resource updated/some message/anything in the current page.
 This function receives the response body as parameter.


## Example

You can see a working example using this component in [here](https://github.com/SeasonedSoftware/croods/blob/master/example/src/colors/Edit.js).
