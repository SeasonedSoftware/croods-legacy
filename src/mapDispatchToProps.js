import { bindActionCreators } from 'redux'

import createActions from './createActions'

export default (dispatch, { options, ...props }) => ({
  actions: bindActionCreators(
    createActions({ ...options, ...props }),
    dispatch,
  ),
})
