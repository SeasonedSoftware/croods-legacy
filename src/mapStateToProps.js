import find from 'lodash/find'
import get from 'lodash/get'

export default (state, { name, parentId }) => {
  const statePart = get(state, name)
  const currentState = find(statePart, block => block.parentId === parentId)
  return get(currentState, 'state') || {}
}
