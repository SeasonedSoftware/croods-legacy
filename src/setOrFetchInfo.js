import find from 'lodash/find'

export default ({ id, info, list, actions }) => {
  const item = find(
    list || [],
    item => item && item.id.toString() === id.toString(),
  )

  if (item) {
    actions.setInfo(item)
  } else if (!info || info.id.toString() !== id.toString()) {
    actions.fetchInfo(id)
  }
}
