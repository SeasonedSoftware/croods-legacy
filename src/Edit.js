import { Component } from 'react'
import { connect } from 'react-redux'

import withOptions from './withOptions'
import mapStateToProps from './mapStateToProps'
import mapDispatchToProps from './mapDispatchToProps'
import setOrFetchInfo from './setOrFetchInfo'

class Edit extends Component {
  constructor(props) {
    super(props)
    setOrFetchInfo(props)
  }

  componentWillReceiveProps(nextProps) {
    const { id } = nextProps
    const { id: oldId } = this.props

    if (id.toString() !== oldId.toString()) {
      setOrFetchInfo(nextProps)
    }
  }

  componentDidUpdate(prevProps) {
    const { actions, updated } = this.props
    const { updated: oldUpdated } = prevProps

    if (updated && !oldUpdated) {
      actions.resetUpdated()
    }
  }

  render() {
    const { render, renderUpdated, actions, info, fetchingInfo } = this.props
    const { infoError, updated, updating, updateError: error } = this.props
    const { renderLoading, renderError } = this.props
    const { update } = actions

    if (infoError) {
      return renderError(infoError)
    }

    if (!info || fetchingInfo) {
      return renderLoading()
    }

    if (updated) {
      return renderUpdated(updated)
    }

    return render({ info, update, updating, error })
  }
}

export default withOptions(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  )(Edit),
)
