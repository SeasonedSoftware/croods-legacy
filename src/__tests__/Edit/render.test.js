import React from 'react'
import { connect } from 'react-redux'
import renderer from 'react-test-renderer'

import Edit from '../../Edit'
import mapStateToProps from '../../mapStateToProps'
import mapDispatchToProps from '../../mapDispatchToProps'
import setOrFetchInfo from '../../setOrFetchInfo'

jest.mock('../../withOptions', () => Component => Component)

jest.mock('react-redux', () => ({
  connect: jest.fn(() => Component => Component),
}))

jest.mock('../../setOrFetchInfo', () => jest.fn())

const render = jest.fn((editProps, props) => (
  <div {...editProps} {...props}>
    Edit
  </div>
))

const actions = { update: jest.fn() }

beforeEach(() => {
  render.mockClear()
  setOrFetchInfo.mockClear()
})

it('renders correctly', () => {
  const props = { id: 1, name: 'colors', render, actions }
  const tree = renderer.create(<Edit {...props} />).toJSON()

  expect(tree).toMatchSnapshot()
  expect(connect).toHaveBeenCalledWith(mapStateToProps, mapDispatchToProps)
  expect(setOrFetchInfo).toHaveBeenCalledWith(props)
})

describe('with renderError and without infoError', () => {
  it('renders edit', () => {
    const renderError = jest.fn(error => <div>Error: {error}</div>)
    const props = { render, actions, renderError }
    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('with renderError and infoError', () => {
  it('renders error', () => {
    const renderError = jest.fn(error => <div>Error: {error}</div>)
    const props = { render, actions, renderError, infoError: 'foo' }
    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('with renderLoading and fetchingInfo', () => {
  it('renders loading correctly', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = { render, actions, fetchingInfo: true, renderLoading }
    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})

describe('with renderLoading and info', () => {
  it('renders edit correctly', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = { render, actions, renderLoading, info: { foo: 'bar' } }
    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})

describe('without renderLoading and with fetchingInfo', () => {
  it('renders edit correctly', () => {
    const props = { render, actions, fetchingInfo: true }
    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).toHaveBeenCalled()
  })
})

describe('with renderLoading, fetchingInfo and info', () => {
  it('renders loading', () => {
    const renderLoading = jest.fn(() => <div>Loading</div>)
    const props = {
      render,
      actions,
      fetchingInfo: true,
      renderLoading,
      info: { foo: 'bar' },
    }

    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})

describe('with renderUpdated and updated', () => {
  it('renders updated', () => {
    const renderUpdated = jest.fn(updated => <div {...updated}>Updated</div>)
    const props = {
      render,
      actions,
      renderUpdated,
      updated: { foo: 'bar' },
    }

    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
    expect(render).not.toHaveBeenCalled()
  })
})

describe('with renderUpdated and not updated', () => {
  it('renders edit', () => {
    const renderUpdated = jest.fn(updated => <div {...updated}>Updated</div>)
    const props = {
      render,
      actions,
      renderUpdated,
    }

    const tree = renderer
      .create(<Edit id={1} name="colors" {...props} />)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
