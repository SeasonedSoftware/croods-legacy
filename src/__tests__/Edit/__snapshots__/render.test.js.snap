// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  actions={
    Object {
      "update": [Function],
    }
  }
  error={undefined}
  id={1}
  info={undefined}
  name="colors"
  render={[Function]}
  update={[Function]}
  updating={undefined}
>
  Edit
</div>
`;

exports[`with renderError and infoError renders error 1`] = `
<div>
  Error: 
  foo
</div>
`;

exports[`with renderError and without infoError renders edit 1`] = `
<div
  actions={
    Object {
      "update": [Function],
    }
  }
  error={undefined}
  id={1}
  info={undefined}
  name="colors"
  render={[Function]}
  renderError={[Function]}
  update={[Function]}
  updating={undefined}
>
  Edit
</div>
`;

exports[`with renderLoading and fetchingInfo renders loading correctly 1`] = `
<div>
  Loading
</div>
`;

exports[`with renderLoading and info renders edit correctly 1`] = `
<div
  actions={
    Object {
      "update": [Function],
    }
  }
  error={undefined}
  id={1}
  info={
    Object {
      "foo": "bar",
    }
  }
  name="colors"
  render={[Function]}
  renderLoading={[Function]}
  update={[Function]}
  updating={undefined}
>
  Edit
</div>
`;

exports[`with renderLoading, fetchingInfo and info renders loading 1`] = `
<div>
  Loading
</div>
`;

exports[`with renderUpdated and not updated renders edit 1`] = `
<div
  actions={
    Object {
      "update": [Function],
    }
  }
  error={undefined}
  id={1}
  info={undefined}
  name="colors"
  render={[Function]}
  renderUpdated={[Function]}
  update={[Function]}
  updating={undefined}
>
  Edit
</div>
`;

exports[`with renderUpdated and updated renders updated 1`] = `
<div
  foo="bar"
>
  Updated
</div>
`;

exports[`without renderLoading and with fetchingInfo renders edit correctly 1`] = `
<div
  actions={
    Object {
      "update": [Function],
    }
  }
  error={undefined}
  fetchingInfo={true}
  id={1}
  info={undefined}
  name="colors"
  render={[Function]}
  update={[Function]}
  updating={undefined}
>
  Edit
</div>
`;
